src/core/term_dag.rs

```diff
@@ -116,6 +115,111 @@ impl<Witness, App: Application> TermDag<Witness, App> {
     pub fn jet(jet: &'static JetNode<App>) -> Rc<Self> {
         Rc::new(TermDag::Jet(jet))
     }
+
+    /// Return the left child of the given DAG root, if there is such a child.
+    pub fn get_left(&self) -> Option<&Self> {
+        match self {
+            TermDag::InjL(l) | TermDag::InjR(l) | TermDag::Take(l) | TermDag::Drop(l) => Some(l),
+            TermDag::Comp(l, _)
+            | TermDag::Case(l, _)
+            | TermDag::Pair(l, _)
+            | TermDag::AssertL(l, _)
+            | TermDag::AssertR(l, _)
+            | TermDag::Disconnect(l, _) => Some(l),
+            _ => None,
+        }
+    }
+
+    /// Return the right child of the given DAG root, if there is such a child.
+    pub fn get_right(&self) -> Option<&Self> {
+        match self {
+            TermDag::Comp(_, r)
+            | TermDag::Case(_, r)
+            | TermDag::Pair(_, r)
+            | TermDag::AssertL(_, r)
+            | TermDag::AssertR(_, r)
+            | TermDag::Disconnect(_, r) => Some(r),
+            _ => None,
+        }
+    }
+
+    /// Visit the DAG and return a stack that can be popped such that the nodes are in post order,
+    /// i.e., children appear before their parent.
+    fn get_post_order_stack(&self) -> Vec<&TermDag<Witness, App>> {
+        let mut visited = Vec::new();
+        let mut to_visit = vec![self];
+
+        while let Some(dag) = to_visit.pop() {
+            visited.push(dag);
+
+            if let Some(l) = dag.get_left() {
+                if !l.is_cross_reference(&to_visit) {
+                    to_visit.push(l);
+                }
+            }
+            if let Some(r) = dag.get_right() {
+                if !r.is_cross_reference(&to_visit) {
+                    to_visit.push(r);
+                }
+            }
+        }
+
+        visited
+    }
+
+    /// Return whether `self` is a cross-reference to a node that appears
+    /// earlier in the post-order of the DAG.
+    ///
+    /// Using the two-stack approach of [`Self::get_post_order_stack`],
+    /// nodes that appear earlier are reachable from the nodes in `to_visit`.
+    fn is_cross_reference(&self, to_visit: &[&TermDag<Witness, App>]) -> bool {
+        for dag in to_visit {
+            if self.is_reachable_from(dag) {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    // TODO: Slow for large programs, like HASHBLOCK or SCHNORR
```

Comment:
In 41ae92f368f222738c1b99058f11f723b86e1155:

Fine to leave this as a TODO for now. (Though arguably it should be a github issue rather than being a code comment, but that's fine.) A couple solutions, for posterity..
* Memoize -- I think this is probably the best thing. When you convert this to an iterator you can just stick a `HashSet<TermDagWrapper>` where the wrapper type is just something that implements `Eq` and `Hash` based on cmr rather than trying to recurse into the object.
* Stick a `RefCell<bool>` into the `TermDag`s themselves and mark each one `true` when it's first visited...though I can't figure out how to make sure they're all set to `false` at the start of iteration so maybe this is a no-go
* The above, but with pointer tagging rather than an explicit `RefCell<bool>` :). I'm pretty sure this is just impossible in Rust.

Reply:
Actually, on further thought I think we need to memoize based on pointer identity (as you're doing now) rather than CMR.

Reply:
Ok. Godspeed.